src/engine/idTable/CompressedExternalIdTable.h

```diff
@@ -209,49 +211,96 @@ class CompressedExternalIdTableWriter {
-    auto firstBlock = startOfSingleIdTables_.at(index);
-    auto lastBlock = index + 1 < startOfSingleIdTables_.size()
-                         ? startOfSingleIdTables_.at(index + 1)
-                         : blocksPerColumn_.at(0).size();
-    std::future<Table> fut;
-
-    // Yield one block after the other. While one block is yielded the next
-    // block is already read concurrently.
-    for (size_t blockIdx = firstBlock; blockIdx < lastBlock; ++blockIdx) {
-      std::optional<Table> table;
-      if (fut.valid()) {
-        table = fut.get();
+  template <size_t NumCols = 0>
+  struct TableBlockGenerator {
+    CompressedExternalIdTableWriter* writer_;
+    size_t index_;
+    size_t firstBlock_;
+    size_t lastBlock_;
+    size_t blockIdx_;
+    std::future<IdTableStatic<NumCols>> fut_;
+    bool isFirstCall_ = true;
+    bool hasDecremented_ = false;
+
+    TableBlockGenerator(CompressedExternalIdTableWriter* writer, size_t index)
+        : writer_(writer),
+          index_(index),
+          firstBlock_(writer->startOfSingleIdTables_.at(index)),
+          lastBlock_(index + 1 < writer->startOfSingleIdTables_.size()
+                         ? writer->startOfSingleIdTables_.at(index + 1)
+                         : writer->blocksPerColumn_.at(0).size()),
+          blockIdx_(firstBlock_) {
+      ++writer_->numActiveGenerators_;
+    }
+    ~TableBlockGenerator() {
+      if (!hasDecremented_) {
+        --writer_->numActiveGenerators_;
       }
-      fut = std::async(std::launch::async, [this, blockIdx]() {
-        return readBlock<NumCols>(blockIdx);
+    }
+
+    std::optional<IdTableStatic<NumCols>> operator()() {
+      absl::Cleanup cleanup([this] {
+        --writer_->numActiveGenerators_;
+        hasDecremented_ = true;
       });
-      if (table.has_value()) {
-        co_yield table.value();
+      if (isFirstCall_) {
+        isFirstCall_ = false;
+        if (blockIdx_ >= lastBlock_) {
+          return std::nullopt;
+        }
+        fut_ = std::async(std::launch::async, [this]() {
+          return this->writer_->template readBlock<NumCols>(this->blockIdx_++);
+        });
+        AD_CONTRACT_CHECK(
+            fut_.valid(),
+            "Future initialization failed in TableBlockGenerator");
+        auto table = fut_.get();
+        if (blockIdx_ < lastBlock_) {
+          fut_ = std::async(std::launch::async, [this]() {
+            return this->writer_->template readBlock<NumCols>(
+                this->blockIdx_++);
+          });
+          std::move(cleanup).Cancel();
+        }
+        return table;
+      }
+      if (blockIdx_ > lastBlock_) {
+        return std::nullopt;
+      }
+      if (!fut_.valid()) {
+        return std::nullopt;
       }
+      auto table = fut_.get();
+      if (blockIdx_ < lastBlock_) {
+        fut_ = std::async(std::launch::async, [this]() {
+          return this->writer_->template readBlock<NumCols>(this->blockIdx_++);
+        });
```

Comment:
The old implementation has a special case for the last block.

Your implementation has a special case for the first AND the last iteration.
This shouldn't be necessary. In particular: 
0. If out of bounds (blockIdx > lastBlock) return nullopt.
1.  Only the first time (the if is okay) start the future (the first table could also computed synchronously, but not important for now).
2.  Get the table from the future (for all blocks).
3. Only if not last block start the next future
4. Always increase the block Idx by 1
5. Return the table.

This workflow should be much easier to follow.